pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinitionSimple.java

```diff
@@ -26,21 +26,38 @@
 
 
 /* default */ class JavaTypeDefinitionSimple extends JavaTypeDefinition {
+
+    static final JavaTypeDefinitionSimple OBJECT_DEFINITION = new JavaTypeDefinitionSimple(Object.class);
     private final Class<?> clazz;
-    private final JavaTypeDefinition[] genericArgs;
+    private JavaTypeDefinition[] genericArgs;
     // cached because calling clazz.getTypeParameters().length create a new array every time
-    private final int typeParameterCount;
-    private final boolean isGeneric;
-    private final boolean isRawType;
-    private final JavaTypeDefinition enclosingClass;
+    private int typeParameterCount = -1;
+    private final int typeArgumentCount;
 
     private static final Logger LOG = Logger.getLogger(JavaTypeDefinitionSimple.class.getName());
-    private static final JavaTypeDefinition[] NO_GENERICS = {};
 
     protected JavaTypeDefinitionSimple(Class<?> clazz, JavaTypeDefinition... boundGenerics) {
         super(EXACT);
         this.clazz = clazz;
 
+        typeArgumentCount = boundGenerics.length;
+        if (boundGenerics.length > 0) {
+            genericArgs = Arrays.copyOf(boundGenerics, boundGenerics.length);
+        } // otherwise stays null
+    }
+
+    private Class<?> loadEnclosing(Class<?> clazz) {
+        try {
+            return clazz.getEnclosingClass();
+        } catch (LinkageError e) {
+            if (LOG.isLoggable(Level.WARNING)) {
+                LOG.log(Level.WARNING, "Could not load enclosing class of " + clazz.getName() + ", due to: " + e);
+            }
+            return null;
+        }
+    }
+
+    private TypeVariable<?>[] getTypeParameters(Class<?> clazz) {
```

Comment:
I think, this method might throw LinkageError when loading the interfaces or superclass.